### PR TITLE
Fix parsing of optional arguments for functions without mandatory args

### PR DESCRIPTION
**Pull request by marmeladema**

### CODE REVIEW

File: engine/src/ast/function_expr.rs [hunk 1]
```diff
@@ -164,8 +164,15 @@ impl<'i, 's> LexWith<'i, &'s Scheme> for FunctionCallExpr<'s> {
 
         let mut index = 0;
 
-        while let Some(',') = input.chars().next() {
-            input = skip_space(take(input, 1)?.1);
+        while let Some(c) = input.chars().next() {
+            if c == ')' {
+                break;
+            }
+            if !function_call.args.is_empty() {
```
Comment (RReverser) [inline engine/src/ast/function_expr.rs]: This seems too specific. If you haven't seen `)`, then you should always expect `,`, no?

Comment (marmeladema) [reply engine/src/ast/function_expr.rs]: Maybe i misunderstood your comment, but in case there are no mandatory arguments and the first argument is an optional one then you should not expect a `,` in this case. Doing so would lead to parse `func(,arg)` as valid.

Comment (RReverser) [reply engine/src/ast/function_expr.rs]: Then, instead, you should check `index` to make it more obvious that you don't expect a `,` before the first argument (when `index == 0`). And move skip_space inside the condition, since otherwise it's not necessary.

File: engine/src/ast/function_expr.rs [hunk 2]
```diff
@@ -164,8 +164,15 @@ impl<'i, 's> LexWith<'i, &'s Scheme> for FunctionCallExpr<'s> {
 
         let mut index = 0;
 
-        while let Some(',') = input.chars().next() {
-            input = skip_space(take(input, 1)?.1);
+        while let Some(c) = input.chars().next() {
+            if c == ')' {
```
Comment (RReverser) [inline engine/src/ast/function_expr.rs]: Overall, this parsing code should be simplified somehow, since it currently duplicates many checks for the same char, making flow harder to read. But it's ok to leave this to a separate PR.

Comment (marmeladema) [reply engine/src/ast/function_expr.rs]: :+1:  I agree, i just did not wanted to refactor this piece in the same PR.

File: engine/src/ast/function_expr.rs (hunk 1 above)
Comment (marmeladema) [reply engine/src/ast/function_expr.rs]: `index  == 0` does only mean `first optional argument`, there could have been zero or more mandatory arguments. We only expect a `,` if the argument is not the first.

Comment (RReverser) [reply engine/src/ast/function_expr.rs]: Oh, okay. This seems easy to miss, can you leave a comment?

Comment (marmeladema) [reply engine/src/ast/function_expr.rs]: Ok, i'll add one ASAP.

Comment (marmeladema) [reply engine/src/ast/function_expr.rs]: Done

**Comment by marmeladema** [comment]
Once this is merged, i'll refactor the parsing of arguments on top of #29
